### ISSUE TITLE
Fail to compiling a static executable vw on macOS

### ISSUE THREAD

**Issue by silverguo** [VowpalWabbit/vowpal_wabbit#1749]
#### Describe the bug
Following the instruction [here](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Building), I fail to build vw statically on macOS

#### To Reproduce
1. Install all dependencies as mentioned [here](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Dependencies#macos), `brew install boost --with-python` doesn't work for me, I just did `brew install boost`

2. Following the instructions
```
mkdir build
cd build
cmake .. -DSTATIC_LINK_VW=ON
make vw-bin -j
```
#### Expected behavior
A static vw

#### Observed Behavior
```
[ 97%] Linking CXX static library libvw.a
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: libvw.a(parse_example_json.cc.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: libvw.a(parse_example_json.cc.o) has no symbols
[ 97%] Built target vw
Scanning dependencies of target vw-bin
[ 98%] Building CXX object vowpalwabbit/CMakeFiles/vw-bin.dir/main.cc.o
[100%] Linking CXX executable vw
ld: library not found for -lcrt0.o
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [vowpalwabbit/vw] Error 1
make[2]: *** [vowpalwabbit/CMakeFiles/vw-bin.dir/all] Error 2
make[1]: *** [vowpalwabbit/CMakeFiles/vw-bin.dir/rule] Error 2
make: *** [vw-bin] Error 2
```

#### Environment
- vw version: clone from master
- OS: macOS Mojave
- cmake: 3.13.4
- boost: 1.68.0

#### Additional context
I try to build static with lastest vw release 8.6.1 (Jul 21, 2018), but I didn't find the clear readme to build static vw on macOS for this version.

**Comment by arielf** [comment]
I believe this is a more general (Linux as well) not just Mac OS-X issue.

I wasn't able to build statically since the switch to `cmake` building was made. Neither with `clang++` nor with `g++`.

Prior to that, building statically was simple (just add `-static` to `CXX`) and always worked for me.

Building statically is very important for more system-independent production deployments.

**Comment by silverguo** [comment]
@arielf 
Thanks for your reply!

Have you successfully built the VW on macOS with latest release 8.6.1? I try to build it statically with `make CXX='clang++ -static'`, but if I check the binary `vw` with `otool -L vowpalwabbit/vw`, I still get following output, 

    /usr/local/opt/boost/lib/libboost_program_options-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.200.5)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.4)

If I send `vw` to another mac, I still need to install boost in that machine.

Do you know how could I get rid of first dependence for boost?

**Comment by arielf** [comment]
@silverguo sorry, I'm behind responding to this.
I don't use Mac OS-X so no (sorry!).

However, I think we are not yet there with clang+static linking even on Ubuntu, I manage to build with clang, but some tests are failing. Ubuntu 18.04.

static linking has been fixed (works with standard gcc), so no need to add `-static` anymore to `CXX`, it should be sufficient to do this to get static linking:
```
make clean
mkdir -p build

( cd build && cmake .. -DCMAKE_BUILD_TYPE=Release -DSTATIC_LINK_VW=ON )

make -j
```

**Comment by silverguo** [comment]
> ```
> make clean
> mkdir -p build
> 
> ( cd build && cmake .. -DCMAKE_BUILD_TYPE=Release -DSTATIC_LINK_VW=ON )
> 
> make -j
> ```

@arielf , I follow these command on Ubuntu, I could build static `vw`, but the `spanning_tree` is not static (`ldd cluster/spanning_tree` will show several dependencies), do you have the same issue?

But with old version (before `cmake`), I could build static `spanning_tree`.

How could we build static `spanning_tree` with `cmake`?

**Comment by jackgerrits** [comment]
Hi @silverguo, I looked into this and `spanning_tree` was not hooking into the same logic for static executable creation. I've created a PR (#1787) that fixes this, no you can run cmake with  `-DSTATIC_LINK_VW=ON` and the `spanning_tree` executable you build should be staticly linked.

**Comment by silverguo** [comment]
@jackgerrits Great, thanks!

**Comment by arielf** [comment]
Thanks @jackgerrits !

**Comment by jackgerrits** [comment]
Looks like @kumpera has addressed this in #1807, if there are still issues let us know
